### PR TITLE
Check the presence of glyphs when selecting font

### PR DESCRIPTION
**Pull request by bwikbs**

### CODE REVIEW

**Comment by bwikbs** [comment]
https://github.com/flutter-tizen/engine/issues/96

**Comment by bwikbs** [comment]
dump_screen
A problem was identified in a specific case. Cause analysis is required.
- Galaxy Watch3, Gallery App

**Comment by bwikbs** [comment]
A patch was created based on the analysis of @bbrto21.(#96) The above situation has been resolved.
But..I'm not sure this is right approche of this. :cry:

File: third_party/txt/src/minikin/FontCollection.cpp [hunk 1]
```diff
@@ -509,7 +509,9 @@ void FontCollection::itemize(const uint16_t* string,
     if (!shouldContinueRun) {
       const std::shared_ptr<FontFamily>& family = getFamilyForChar(
           ch, isVariationSelector(nextCh) ? nextCh : 0, langListId, variant);
-      if (utf16Pos == 0 || family.get() != lastFamily) {
+
+      if (utf16Pos == 0 || family.get() != lastFamily ||
+          family.get()->getLastMatchedCodePoint() != ch) {
```
Comment (bbrto21) [inline third_party/txt/src/minikin/FontCollection.cpp:514]: I think that using only one codepoint causes too much execution.

Comment (bbrto21) [inline third_party/txt/src/minikin/FontCollection.cpp:514]: This algorithm tries to reuse the already selected font as much as possible.
I think that using only one codepoint cached causes too much execution.

File: third_party/txt/src/minikin/FontCollection.cpp [hunk 2]
```diff
@@ -528,8 +530,8 @@ void FontCollection::itemize(const uint16_t* string,
           }
           start -= prevChLength;
         }
-        result->push_back(
-            {family->getClosestMatch(style), static_cast<int>(start), 0});
+        result->push_back({family->getClosestMatchWithChar(style, ch),
```
Comment (bbrto21) [inline third_party/txt/src/minikin/FontCollection.cpp:533]: Shouldn't we also use variant to check glyphs?
```C++
result = hb_font_get_glyph(hb_font, codepoint, variant, &unusedGlyph);
```

File: third_party/txt/src/minikin/FontFamily.cpp [hunk 3]
```diff
@@ -164,6 +164,46 @@ FakedFont FontFamily::getClosestMatch(FontStyle style) const {
   return FakedFont{nullptr, FontFakery()};
 }
 
+FakedFont FontFamily::getClosestMatchWithChar(FontStyle style,
```
Comment (bbrto21) [inline third_party/txt/src/minikin/FontFamily.cpp:167]: The core algorithm is similar to ```FakedFont FontFamily::getClosestMatch(FontStyle style)```
I think you can make one nice method for these two.

File: third_party/txt/src/txt/platform_linux.cc [hunk 4]
```diff
@@ -13,6 +13,9 @@
 namespace txt {
 
 std::vector<std::string> GetDefaultFontFamilies() {
+#ifdef FLUTTER_USE_FONTCONFIG
```
Comment (bbrto21) [inline third_party/txt/src/txt/platform_linux.cc:16]: IMHO, if we fix this issue properly, if so, the previously committed code should be revert.
- 6e859d7ce (tag: 6e859d7) Create fallback font manager to solve performance drops (#78)
- 28dbd0208 (plugin_registrar) Fix font breaking issues (#69)

and I think we should only have one to two default fonts such as ```SamsungOneUI``` or ```TizenDefaultFont``` like other platform implementations

**Review by bbrto21 — COMMENTED** [review]
I think you need to re-enable fontconfig in azure-pipelines.

File: third_party/txt/src/minikin/FontCollection.cpp (hunk 1 above)
Comment (bwikbs) [reply third_party/txt/src/minikin/FontCollection.cpp:514]: You're right. Maybe I should use map for this.

File: third_party/txt/src/minikin/FontCollection.cpp (hunk 2 above)
Comment (bwikbs) [reply third_party/txt/src/minikin/FontCollection.cpp:533]: I didn't know about ``variant selector``, so I looked it up. I think you're right. :smile:

File: third_party/txt/src/minikin/FontFamily.cpp (hunk 3 above)
Comment (bwikbs) [reply third_party/txt/src/minikin/FontFamily.cpp:167]: Ok~ Thanks!

**Comment by bbrto21** [comment]
And it still have the similar problem as the image attached below.

dump_screen

File: third_party/txt/src/txt/platform_linux.cc (hunk 4 above)
Comment (bwikbs) [reply third_party/txt/src/txt/platform_linux.cc:16]: I agree with you. 
If this patch lands, can you do it next time? :smile:

**Comment by bwikbs** [comment]
> And it still have the similar problem as the image attached below.
> 
> dump_screen

OMG! It's not easy...

**Comment by bbrto21** [comment]
I've made some progress based on your work, I think it's almost done. there is room for improvement, so please refer to this in your implementation. 
https://github.com/bbrto21/engine/commit/563b60c70905ce12d67f5d659d3425af21b09d82 (I worked on it after revert 6e859d7, 28dbd02)

dump_screen-1
dump_screen-2
dump_screen-3
dump_screen-4

**Comment by bwikbs** [comment]
@bbrto21 I put your implementation as a patch. As you said, there seems to be no problem. Thanks. :+1: 
Reverting existing code and applying fontconfig seem to be a little different topic, so I would like to talk about it in another PR.

File: third_party/txt/src/minikin/FontCollection.cpp [hunk 5]
```diff
@@ -509,6 +509,7 @@ void FontCollection::itemize(const uint16_t* string,
     if (!shouldContinueRun) {
       const std::shared_ptr<FontFamily>& family = getFamilyForChar(
           ch, isVariationSelector(nextCh) ? nextCh : 0, langListId, variant);
+
```
Comment (bbrto21) [inline third_party/txt/src/minikin/FontCollection.cpp:512]: :scissors:

File: third_party/txt/src/minikin/FontCollection.cpp [hunk 6]
```diff
@@ -528,8 +529,27 @@ void FontCollection::itemize(const uint16_t* string,
...
-        result->push_back(
-            {family->getClosestMatch(style), static_cast<int>(start), 0});
+        result->push_back({family->getClosestMatch(style, ch, variant),
+                           static_cast<int>(start), 0});
+        run = &result->back();
+        lastFamily = family.get();
+      } else if (family.get() == lastFamily &&
+                 // We need to change this condition to more general logic. And
+                 // when changing the font, it is necessary to check whether the
+                 // existing characters also have glyphs.
```
Comment (bbrto21) [inline third_party/txt/src/minikin/FontCollection.cpp:539]: Yes, I agree too and I think font class must provide a function to check the presence of glyphs.

File: third_party/txt/src/minikin/FontFamily.cpp [hunk 7]
```diff
@@ -146,15 +146,27 @@ static FontFakery computeFakery(FontStyle wanted, FontStyle actual) {
   return FontFakery(isFakeBold, isFakeItalic);
 }
 
-FakedFont FontFamily::getClosestMatch(FontStyle style) const {
+FakedFont FontFamily::getClosestMatch(FontStyle style,
+                                      uint32_t codepoint,
+                                      uint32_t variationSelector) const {
```
Comment (bbrto21) [inline third_party/txt/src/minikin/FontFamily.cpp:151]: I recommend this for readability, but you can ignore it.

```suggestion
                                      uint32_t codepoint /* = 0 */,
                                      uint32_t variationSelector /* = 0 */) const {
```

File: third_party/txt/src/minikin/FontFamily.cpp [hunk 8]
```diff
@@ -146,15 +146,27 @@ static FontFakery computeFakery(FontStyle wanted, FontStyle actual) {
...
   for (size_t i = 0; i < mFonts.size(); i++) {
     const Font& font = mFonts[i];
     int match = computeMatch(font.style, style);
-    if (i == 0 || match < bestMatch) {
-      bestFont = &font;
-      bestMatch = match;
+    bool result = false;
+    if (codepoint != 0) {
+      hb_font_t* hb_font = getHbFontLocked(font.typeface.get());
+      uint32_t unusedGlyph;
```
Comment (bbrto21) [inline third_party/txt/src/minikin/FontFamily.cpp:160]:
```suggestion
      uint32_t unusedGlyph = 0;
```

File: third_party/txt/src/txt/platform_linux.cc (hunk 4 above)
Comment (bbrto21) [reply third_party/txt/src/txt/platform_linux.cc:16]: Ok

**Comment by bbrto21** [comment]
@bwikbs @HakkyuKim 
According to further investigation, unfortunately, SamsungOneUI font family has color emoji font. 
It is sometimes chosen for non-emoji characters. Especially when rendering text with only numbers it is always selected.

**Comment by bwikbs** [comment]
No matter how much I think about it, I can't find an answer, so I made a workaround that the emoji goes backwards when sorting. 
@bbrto21 @HakkyuKim How do you think?  It works in all cases as we found before.

**Comment by bwikbs** [comment]
Go on https://github.com/flutter-tizen/engine/pull/113

**Pull request closed without merging**